File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-13.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<int[]> stack = new Stack();
        int maxArea = 0;
        for (int i = 0; i < heights.length; i++){
            int curr = heights[i];
            int prevPos = i;
            while (!stack.empty() && stack.peek()[0] >= curr){
                int[] val = stack.pop();
                prevPos = val[1];
                int area = val[0]* (i-val[1]);
                maxArea = Math.max(maxArea, area);
            }

            stack.push(new int[]{curr, prevPos});
        }
        while (!stack.empty()){
            int[] val = stack.pop();
            int area = val[0] * (heights.length-val[1]);
            maxArea = Math.max(maxArea, area);
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-13.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        L, R = 0, len(heights)-1
        max_left, max_right = heights[0], heights[-1]
        total = 0

        while L < R:
            if max_left < max_right:
                L += 1
                max_left = max(max_left, heights[L])
                total += max_left - heights[L]
            else:
                R -= 1
                max_right = max(max_right, heights[R])
                total += max_right - heights[R]
        return total